src/middlewared/middlewared/plugins/cloud_sync.py

```diff
@@ -0,0 +1,522 @@
+from middlewared.service import (
+    CallError, CRUDService, ValidationErrors, item_method, filterable, job, private
+)
+from middlewared.utils import load_modules, load_classes, Popen, run
+
+import asyncio
+import base64
+import codecs
+from collections import namedtuple
+from Crypto import Random
+from Crypto.Cipher import AES
+from Crypto.Util import Counter
+import json
+import os
+import subprocess
+import re
+import tempfile
+
+CHUNK_SIZE = 5 * 1024 * 1024
+
+REMOTES = {}
+
+RcloneConfigTuple = namedtuple("RcloneConfigTuple", ["config_path", "remote_path"])
+
+
+class RcloneConfig:
+    def __init__(self, cloud_sync):
+        self.cloud_sync = cloud_sync
+
+        self.provider = REMOTES[self.cloud_sync["credentials"]["provider"]]
+
+        self.tmp_file = None
+        self.path = None
+
+    def __enter__(self):
+        self.tmp_file = tempfile.NamedTemporaryFile(mode='w+')
+
+        # Make sure only root can read it as there is sensitive data
+        os.chmod(self.tmp_file.name, 0o600)
+
+        config = dict(self.cloud_sync["credentials"]["attributes"], type=self.provider.rclone_type)
+        config = dict(config, **self.provider.get_credentials_extra(self.cloud_sync["credentials"]))
+        if "pass" in config:
+            config["pass"] = rclone_encrypt_password(config["pass"])
+
+        remote_path = None
+
+        if "attributes" in self.cloud_sync:
+            config.update(dict(self.cloud_sync["attributes"], **self.provider.get_task_extra(self.cloud_sync)))
+
+            remote_path = "remote:" + "/".join([self.cloud_sync["attributes"].get("bucket", ""),
+                                                self.cloud_sync["attributes"].get("folder", "")]).strip("/")
+
+            if self.cloud_sync.get("encryption"):
+                self.tmp_file.write("[encrypted]\n")
+                self.tmp_file.write("type = crypt\n")
+                self.tmp_file.write("remote = %s\n" % remote_path)
+                self.tmp_file.write("filename_encryption = %s\n" % ("standard" if self.cloud_sync["filename_encryption"]
+                                                                    else "off"))
+                self.tmp_file.write("password = %s\n" % rclone_encrypt_password(self.cloud_sync["encryption_password"]))
+                self.tmp_file.write("password2 = %s\n" % rclone_encrypt_password(self.cloud_sync["encryption_salt"]))
+
+                remote_path = "encrypted:/"
+
+        self.tmp_file.write("[remote]\n")
+        for k, v in config.items():
+            self.tmp_file.write(f"{k} = {v}\n")
+
+        self.tmp_file.flush()
+
+        return RcloneConfigTuple(self.tmp_file.name, remote_path)
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self.tmp_file:
+            self.tmp_file.close()
+
+
+async def rclone(job, cloud_sync):
+    # Use a temporary file to store rclone file
+    with RcloneConfig(cloud_sync) as config:
+        args = [
+            '/usr/local/bin/rclone',
```

Comment:
The mix of single and double quotes is messing with my OCD = P. I know its annoying and not important, but can we pick one for each plugin? If a plugin uses double quote use it for all in the same plugin. If it uses single quote, choose single quote for everything in that plugin.